src/client.ts

```diff
@@ -236,6 +236,21 @@ const everyTagPredicate = (tags: string | string[]): string =>
 const someTagsPredicate = (tags: string | string[]): string =>
 	predicate.any("document.tags", castArray(tags));
 
+/**
+ * Type definitions for the `createClient()` function. May be augmented by
+ * third-party libraries.
+ */
+export interface CreateClient {
+	<
+		DocumentTypeMap extends Record<string, prismicT.PrismicDocument> = Record<
+			string,
+			prismicT.PrismicDocument
+		>,
+	>(
+		...args: ConstructorParameters<typeof Client>
+	): Client<DocumentTypeMap>;
+}
```

Comment:
This is needed to pair a repository name literal to its document types. This allows this new system to support apps with multiple repositories, each with their own set of document types.

For example:

```ts
declare module "@prismicio/client" {
  interface CreateClient {
    (
      repositoryNameOrEndpoint: "foo",
      options?: prismic.ClientConfig | undefined
    ): prismic.Client<FooDocumentTypes>;
    (
      repositoryNameOrEndpoint: "bar",
      options?: prismic.ClientConfig | undefined
    ): prismic.Client<BarDocumentTypes>;
  }
}
```

While this is a rare situation, it ensure we are not locked into only supporting a single global config.

Since this code would typically be generated rather than hand-written, the extra boilerplate is okay, I think.

Reply:
OK, I was thinking it'll be something like that, great!